RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc

```diff
@@ -398,47 +803,107 @@ namespace {
-  }
+  TransientTrackingRecHit::ConstRecHitContainer output;
+  bestHit = MTDHitMatchingInfo();
+  for (const DetLayer* ilay : layers)
+    fillMatchingHits(ilay, tsos, traj, hits, prop, bs, vtxTime, matchVertex, output, bestHit);
   return output;
 }
 
 template <class TrackCollection>
 TransientTrackingRecHit::ConstRecHitContainer TrackExtenderWithMTDT<TrackCollection>::tryETLLayers(
     const TrackType& track,
+    const Trajectory& traj,
     const MTDTrackingDetSetVector& hits,
     const MTDDetLayerGeometry* geo,
     const MagneticField* field,
-    const Propagator* prop) const {
-  TransientTrackingRecHit::ConstRecHitContainer output;
+    const Propagator* prop,
+    const reco::BeamSpot& bs,
+    const double& vtxTime,
+    const bool matchVertex,
+    MTDHitMatchingInfo& bestHit) const {
   const vector<const DetLayer*>& layers = geo->allETLLayers();
 
   auto tTrack = builder->build(track);
+  // get the outermost trajectory point on the track
+  TrajectoryStateOnSurface tsos = tTrack.outermostMeasurementState();
 
+  TransientTrackingRecHit::ConstRecHitContainer output;
+  bestHit = MTDHitMatchingInfo();
   for (const DetLayer* ilay : layers) {
     const BoundDisk& disk = static_cast<const MTDRingForwardDoubleLayer*>(ilay)->specificSurface();
     const double diskZ = disk.position().z();
 
-    // get the outermost trajectory point on the track
-    TrajectoryStateOnSurface tsos = tTrack.outermostMeasurementState();
     if (tsos.globalPosition().z() * diskZ < 0)
       continue;  // only propagate to the disk that's on the same side
-    find_hits_in_dets(hits, ilay, tsos, prop, *theEstimator, *hitbuilder, output);
+
+    fillMatchingHits(ilay, tsos, traj, hits, prop, bs, vtxTime, matchVertex, output, bestHit);
   }
   return output;
 }
 
+template <class TrackCollection>
+void TrackExtenderWithMTDT<TrackCollection>::fillMatchingHits(const DetLayer* ilay,
+                                                              const TrajectoryStateOnSurface& tsos,
+                                                              const Trajectory& traj,
+                                                              const MTDTrackingDetSetVector& hits,
+                                                              const Propagator* prop,
+                                                              const reco::BeamSpot& bs,
+                                                              const double& vtxTime,
+                                                              const bool matchVertex,
+                                                              TransientTrackingRecHit::ConstRecHitContainer& output,
+                                                              MTDHitMatchingInfo& bestHit) const {
+  std::set<MTDHitMatchingInfo> hitsInLayer;
+  bool hitMatched = false;
+
+  if (useVertex_ && matchVertex)
+    find_hits_in_dets(hits, traj, ilay, tsos, vtxTime, bs, prop, *theEstimator, *hitbuilder, true, hitsInLayer);
```

Comment:
most parameters are the same in the 3 calls here.  std::bind may make this more readable
```c++
using namespace std::placeholders;
short_hits_in_dets = std::bind(find_hits_in_dets, hits, traj, ilay, tsos, _1, bs, prop, *theEstimator, *hitbuilder, _2, hitsInLayer);
...
if (useVertex_ && matchVertex)
   short_hits_in_dets(vtxTime, true);
else 
   short_hits_in_dets(0, false);
...
short_hits_in_dets(0, false);
```